client/app/actions/ManageAvailabilityActions.js

```diff
@@ -17,31 +18,74 @@ const changeVisibleMonth = (day) => ({
   payload: day,
 });
 
-export const dataLoaded = (slots) => ({
+export const dataLoaded = (slots, loadedMonths) => ({
   type: actionTypes.DATA_LOADED,
-  payload: slots,
+  payload: slots.merge({ loadedMonths }),
 });
 
+/**
+   Number of extra months to preload.
+   0 mean no preloading.
+ */
+const PRELOAD_MONTHS = 2;
+
+/**
+   Given a `date` and number of months to preload, return a List of
+   moment dates representing the start of the month that should be
+   loaded.
+
+   @param {moment} date - Date representing the month
+   @params {number} preloadMonths - Number of months to preload
+   @return {List[moment]} List of months that should be loaded
+ */
+const loadRange = (date, preloadMonths) => {
+  const startOfMonth = date.clone().startOf('month');
+  const endOfMonthExclusive = startOfMonth.clone().add(1, 'month');
+
+  const start = startOfMonth.clone().subtract(preloadMonths, 'months');
+  const end = endOfMonthExclusive.clone().add(preloadMonths, 'months');
+
+  return expandRange(start, end, 'months');
+};
+
+/**
+   Given a list of `months` and a Set of already `loadedMonths`,
+   return an object { start, end } where already loaded months is removed.
+ */
+const removeLoadedMonths = (months, loadedMonths) => ({
+  start: months.find((s) => !loadedMonths.includes(s)),
+  end: months.reverse().find((e) => !loadedMonths.includes(e)),
+});
+
+const monthsToLoad = (day, loadedMonths, preloadMonths) =>
+  removeLoadedMonths(loadRange(day, preloadMonths), loadedMonths);
+
 export const changeMonth = (day) =>
   (dispatch, getState) => {
     dispatch(changeVisibleMonth(day));
 
     const state = getState().manageAvailability;
 
-    harmony.get('/bookables/show', {
-      refId: state.get('listingUuid'),
-      marketplaceId: state.get('marketplaceUuid'),
-      include: ['blocks', 'bookings'].join(','),
-    }).then((response) => {
-      const groups = response.get(':included').groupBy((v) => v.get(':type'));
+    const { start, end } = monthsToLoad(day, state.get('loadedMonths'), PRELOAD_MONTHS);
 
-      const slots = new Map({
-        blocks: groups.get(':block', new List()),
-        bookings: groups.get(':booking', new List()),
-      });
+    if (start && end) {
+      harmony.get('/bookables/show', {
+        refId: state.get('listingUuid'),
+        marketplaceId: state.get('marketplaceUuid'),
+        include: ['blocks', 'bookings'].join(','),
+        start: start.toJSON(),
+        end: end.toJSON(),
+      }).then((response) => {
+        const groups = response.get(':included').groupBy((v) => v.get(':type'));
 
-      dispatch(dataLoaded(slots));
-    });
+        const slots = Immutable.Map({
+          blocks: groups.get(':block', Immutable.List()),
+          bookings: groups.get(':booking', Immutable.List()),
+        });
```

Comment:
That's what was discussed in the chat. We also seem to have an exception set for `Immutable.List` etc. that they don't need to have the `new`: https://github.com/sharetribe/sharetribe/blob/master/client/.eslintrc.js#L161